### PR TITLE
Packages: Move animation to new structure

### LINKED ISSUE #6042: Packages: Move `animation` to new structure
## Feature Description

* [ ] Move code from `/assets/src/animation` to `/packages/animation/src/`
* [ ] Update imports where necessary.  
  Use `@web-stories-wp/...` imports for anything that is already a package, relative imports for everything else.
* [ ] Create `/packages/animation/package.json` referencing its dependencies
* [ ] Create `/packages/animation/README.md` with short documentation.

### Blockers ⚠️ 

Right now `assets/src/animation` has a one dependency on the editor:

* `assets/src/edit-story/types` - for story element prop types

At the same time, the editor includes a bunch of code from `animation`.

We first need to resolve this circular dependency before `animation` can be  become its own package.

Possible solution:

* Create a new `@web-stories-wp/elements` package that contains elements-related utils, JSDoc types, and prop types. Elements themselves could be in the same package, or a new `element-library` package.

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Packages: Move animation to new structure

### QA

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [x] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #6042

### CODE REVIEW

File: packages/animation/src/components/provider.js
Comment (spacedmonkey) [inline packages/animation/src/components/provider.js]: This would be fixed when design system package is in place.

File: packages/animation/src/components/provider.js
Comment (spacedmonkey) [inline packages/animation/src/components/provider.js]: Not sure what to do about this one.

File: packages/animation/src/components/useStoryAnimationContext.js
Comment (spacedmonkey) [inline packages/animation/src/components/useStoryAnimationContext.js]: This would be fixed when design style is in place.

File: packages/animation/src/utils/getGlobalSpace.js
Comment (spacedmonkey) [inline packages/animation/src/utils/getGlobalSpace.js]: @swissspidy Not sure what to do about this one.

File: packages/animation/src/storybookUtils/index.js
Comment (spacedmonkey) [inline packages/animation/src/storybookUtils/index.js]: These are a blocker. These utils can not simple be copied over as they relay on theme. Have any thoughts @swissspidy ?

**Comment by swissspidy** [comment]
Since there are still many blockers, let‘s keep this ticket on hold until they are resolved.

**Comment by spacedmonkey** [comment]
Once #8087 is merge, many of the blocks here are cleared. 

Consider that there are only 3 blocking issues here, I think we can solve them and this one across the line. CC @swissspidy

File: packages/animation/src/storybookUtils/index.js
Comment (spacedmonkey) [reply packages/animation/src/storybookUtils/index.js]: Fixed!

File: assets/src/dashboard/app/views/templateDetails/index.js
Comment (swissspidy) [inline assets/src/dashboard/app/views/templateDetails/index.js]: I think this `clamp` util should not be imported from the animation package.

Let's instead _copy_ it to `dashboard/utils`

File: assets/src/dashboard/components/cardGridItem/cardPreview.js
Comment (swissspidy) [inline assets/src/dashboard/components/cardGridItem/cardPreview.js]: Ditto here, let's not use `clamp` from animation package

File: packages/animation/src/storybookUtils/index.js
Comment (swissspidy) [reply packages/animation/src/storybookUtils/index.js]: Actually I don't think we need `CustomCSS` here _**at all**_. We could just remove it entirely.

@littlemilkstudio for confirming

File: packages/animation/package.json
Comment (swissspidy) [inline packages/animation/package.json:14]: This is the dependency I'd want to avoid :/ The animation package should not depend on the design system just for using `createContext`. I doubt this will be treeshakeable.

We might need the dedicated `@web-stories-wp/react` package sooner or later...

Comment (spacedmonkey) [reply packages/animation/package.json:14]: Yeah, this is a problem elsewhere in the code as well. I think this should be happened elsewhere.

File: assets/src/dashboard/components/cardGridItem/cardPreview.js
Comment (spacedmonkey) [reply assets/src/dashboard/components/cardGridItem/cardPreview.js]: Clamp is also used in edit-story.

File: packages/animation/package.json
Comment (swissspidy) [reply packages/animation/package.json:14]: I mentioned this issue in https://docs.google.com/document/d/1ImRe9wYKWGCAZiq3jro5KLpSJK_NMKXHC6FWqs3nwl0/edit

Seems reasonable to create a dedicated package for this. I'll open a ticket.

File: packages/animation/src/utils/getGlobalSpace.js
Comment (swissspidy) [reply packages/animation/src/utils/getGlobalSpace.js]: As mentioned on the ticket, ideally we'd import these types from a dedicated package.

For the time being, we could just _copy_ the whole type def from `edit/story/types.js` to this file. Adding a `TODO` comment indicating that this should be improved in the future.

File: packages/animation/src/components/provider.js
Comment (swissspidy) [reply packages/animation/src/components/provider.js]: As mentioned on the ticket, ideally we'd import these types from a dedicated package.

For the time being, we could just _copy_ the whole proptype from `edit/story/types.js` to this file. Adding a `TODO` comment indicating that this should be improved in the future.

e.g.

```js
const StoryElementPropType = PropTypes.shape({
  id: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  flip: StoryPropTypes.flip,
  rotationAngle: PropTypes.number.isRequired,
  mask: StoryPropTypes.mask,
  link: StoryPropTypes.link,
  opacity: PropTypes.number,
  lockAspectRatio: PropTypes.bool,
});
```

File: assets/src/dashboard/components/cardGridItem/cardPreview.js
Comment (swissspidy) [reply assets/src/dashboard/components/cardGridItem/cardPreview.js]: Yeah `edit-story` has its own `clamp` util but that's OK. This can be duplicated. It's just a one-liner function.

File: packages/animation/package.json
Comment (swissspidy) [reply packages/animation/package.json:14]: Ticket created: #8159

File: assets/src/dashboard/components/cardGridItem/cardPreview.js
Comment (spacedmonkey) [reply assets/src/dashboard/components/cardGridItem/cardPreview.js]: Do the two clamp functions do the same thing?

Comment (swissspidy) [reply assets/src/dashboard/components/cardGridItem/cardPreview.js]: Yeah pretty much, just with different function signature. But it's ok for those two to be separate.

File: assets/src/dashboard/app/views/templateDetails/index.js
Comment (spacedmonkey) [reply assets/src/dashboard/app/views/templateDetails/index.js]: @BrittanyIRL Any thoughts on this?

File: assets/src/dashboard/components/cardGridItem/cardPreview.js
Comment (spacedmonkey) [reply assets/src/dashboard/components/cardGridItem/cardPreview.js]: So should we use the one from animation package or from edit story?

Comment (swissspidy) [reply assets/src/dashboard/components/cardGridItem/cardPreview.js]: Let's copy the one from the animation package to dashboard.

Leaving edit-story as-is

Comment (spacedmonkey) [reply assets/src/dashboard/components/cardGridItem/cardPreview.js]: We would still need animation as a dependency of dashboard because of `STORY_ANIMATION_STATE`.

Screenshot 2021-07-05 at 13 57 02

Comment (swissspidy) [reply assets/src/dashboard/components/cardGridItem/cardPreview.js]: That's OK. Can be dealt with once we extract story preview into its own package

File: packages/animation/src/components/provider.js
Comment (spacedmonkey) [reply packages/animation/src/components/provider.js]: Fixed

File: packages/animation/src/utils/getGlobalSpace.js
Comment (swissspidy) [inline packages/animation/src/utils/getGlobalSpace.js]: It's not possible to import a prop type (an actual JS object) as a a JSDoc type.

We need to import the actual JSDoc type definition.

That means copying this piece of code:

https://github.com/google/web-stories-wp/blob/270cc1726bce0e25c0ca8e551ed13e66202dc232/assets/src/edit-story/types.js#L245-L262

to `packages/animation/src/types.js` and changing this line here accordingly.

```suggestion
 * @typedef {import('../types').Element} Element
```

File: packages/animation/src/types.js
Comment (swissspidy) [inline packages/animation/src/types.js]: Looks like the `Resource` type def needs to be copied as well. Or just replace with `Object` here, should be OK.

```suggestion
 * @property {Object} resource The element's resource object
```

File: packages/animation/src/storybookUtils/index.js
Comment (maxyinger) [reply packages/animation/src/storybookUtils/index.js]: I think we just use this for documenting the amp portion of the animation parts in storybook, it's not really needed for the animations per se.

So if `CustomCSS` isn't needed for the story preview, then it shouldn't be needed here.

If we really need to, we could just leave these stories & story utils in the root `src/animation` directory or something and just have it be there for storybook documentation/implementations that touch everything else

Comment (maxyinger) [reply packages/animation/src/storybookUtils/index.js]: just tested out the storybook implementation for the AMP portions of the animation parts and they seem to work fine without the `CustomCSS` in here at all, so I see no issues with just removing it entirely

File: assets/src/dashboard/utils/useTemplateView.js
Comment (swissspidy) [inline assets/src/dashboard/utils/useTemplateView.js]: That import looks incorrect. Shouldn't that be

```suggestion
import clamp from './clamp';
```

**Comment by swissspidy** [comment]
@googlebot I consent

**Review by swissspidy — APPROVED** [review]

**Pull request merged**
